rpc/httprpc.c

```diff
@@ -0,0 +1,473 @@
+//
+//  httprpc.c
+//  xdag
+//
+//  Created by Peter Liu on 16/06/20.
+//  Copyright  2018 xrdavies. All rights reserved.
+//
+
+/***********************************************
+*
+*                    include files
+*
+************************************************/
+#if !defined(_WIN32) && !defined(_WIN64)
+#include <sys/types.h>
+#include <sys/un.h>
+#include <unistd.h>
+#include <signal.h>
+#include <errno.h>
+#include <string.h>
+#endif
+#include <sys/stat.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <pthread.h>
+
+#include <event2/event.h>
+#include <event2/http.h>
+#include <event2/buffer.h>
+#include <event2/util.h>
+#include <event2/keyvalq_struct.h>
+#include <event2/http_struct.h>
+
+#include <openssl/evp.h>
+#include <openssl/md5.h>
+#include <openssl/hmac.h>
+
+#include "base64.h"
+
+#include "lst.h"
+#include "log.h"
+#include "../json-rpc/rpc_service.h"
+#include "httprpc.h"
+
+
+/***********************************************
+*
+*                    defines
+*
+************************************************/
+#define HTTP_RPC_VARIBLES_LEN            256
+#define HTTP_RPC_ADDR_LEN                64
+#define HTTP_RPC_WHITE_MAX               8
+#define HTTP_RPC_AUTH_LEN                (HTTP_RPC_VARIBLES_LEN*4)
+
+/***********************************************
+*
+*                    structs and global variables
+*
+************************************************/
+
+//! HTTP status codes
+enum HTTP_RPC_STATUS_CODE
+{
+    HTTPRPC_OK                    			  = 200,
+    HTTPRPC_BAD_REQUEST                   = 400,
+    HTTPRPC_UNAUTHORIZED                  = 401,
+    HTTPRPC_FORBIDDEN                     = 403,
+    HTTPRPC_NOT_FOUND                     = 404,
+    HTTPRPC_BAD_METHOD                    = 405,
+    HTTPRPC_INTERNAL_SERVER_ERROR         = 500,
+    HTTPRPC_SERVICE_UNAVAILABLE           = 503,
+};
+
+char g_httprpc_user[HTTP_RPC_VARIBLES_LEN];
+char g_httprpc_password[HTTP_RPC_VARIBLES_LEN];
+
+char g_httprpc_addr[HTTP_RPC_ADDR_LEN] = {"0.0.0.0"};
+int g_httprpc_port = 7678;
+
+LIST g_httproc_white_host;
+
+typedef struct {
+  NODE node;
+  char addr[HTTP_RPC_ADDR_LEN];
+}rpc_white_host;
+
+/***********************************************
+*
+*                    functions
+*
+************************************************/
+
+static int rpc_host_addr_check_ipv4(const char *host){
+
+  int n[4];
+  char c[4];
+
+  if (!host){
+    return 0;
+  }
+
+  if (sscanf(host, "%d%c%d%c%d%c%d%c",
+             &n[0], &c[0], &n[1], &c[1],
+             &n[2], &c[2], &n[3], &c[3]) == 7)
+
+  {
+    int i;
+    for(i = 0; i < 3; ++i)

```

Comment:
Better use brackets for `for` construction. Currently this part of code demands additional efforts to understand control-flow.